File: Python/Medium/11_Container_With_Most_Water.py
# https://leetcode.com/problems/container-with-most-water/
from typing import  List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        first_idx = 0
        last_idx = len(height)-1
        most_water = 0

        while first_idx<last_idx:
            gap = last_idx - first_idx

            if height[last_idx] < height[first_idx]:
                water =  gap * height[last_idx]
                last_idx -= 1
            else:
                water = gap * height[first_idx]
                first_idx += 1

            if most_water<water:
                most_water = water

        return most_water

print(Solution().maxArea([1,1]))

File: C#/Solutions/Medium/11_Container_With_Most_Water.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Solutions.Medium
{
  internal class _11_Container_With_Most_Water
  {
    public static int maxArea(int[] height)
    {
      int mostWater=0, water=0, firstIdx = 0;
      int lastIdx = height.Length - 1;
      while (firstIdx < lastIdx)
      {
        int gap = lastIdx - firstIdx;
        if (height[firstIdx] < height[lastIdx])
        {
          water = gap * height[firstIdx++]; // firstIdx is incremented after the current value is used with height. Equivalent to height[firstIdx]; firstIdx++;
        }
        else
        {
          water = gap * height[lastIdx--];
        }

        if(mostWater < water)
        {
          mostWater = water;
        }

      }
      return mostWater;
    }
  }
}
